### PR TITLE
feat(cache/pam): Save user's last used broker to the cache

### PR DESCRIPTION
**Pull request by denisonbarbosa**
The default broker for the user used to be stored in memory only. This meant that, on each reboot, the user had to reselect the broker again. Now, the last broker used for authentication by the user is also stored in the database. This way, we can restore it every time.

UDENG-1649

### CODE REVIEW

**Comment by jibel** [comment]
@denisonbarbosa can you rebase before submitting.
Also, I didn't do a thourough review but, for the key, can you use ids instead of names?

File: internal/cache/db.go [hunk 1]
```diff
@@ -32,6 +32,7 @@ const (
 	groupByIDBucketName    = "GroupByID"
 	userToGroupsBucketName = "UserToGroups"
 	groupToUsersBucketName = "GroupToUsers"
+	userToBrokerBucketName = "UserToBroker"
```
[the remark this answers is not in the document]
Comment (denisonbarbosa) [reply internal/cache/db.go:35]: I used UserToBroker to keep it consistent with the other buckets that use ID as well, such as `userToGroups` and `groupsToUsers`. If you look at the database files that we use for the tests, you'll see:
```yaml
GroupToUsers:
  "11111": '{"GID":11111,"UIDs":[1111]}'
  "22222": '{"GID":22222,"UIDs":[2222]}'
  "33333": '{"GID":33333,"UIDs":[3333]}'
  "99999": '{"GID":99999,"UIDs":[2222,3333]}'
  # ...
  UserToGroups:
  "1111": '{"UID":1111,"GIDs":[11111]}'
  "2222": '{"UID":2222,"GIDs":[22222,99999]}'
  "3333": '{"UID":3333,"GIDs":[33333,99999]}'
UserToBroker:
  "1111": '"ExampleBrokerID"'
  "2222": '"ExampleBrokerID"'
  "3333": '"ExampleBrokerID"'
```
IMHO, we should either keep it as `UserToBroker` or rename the others as well.

File: internal/cache/delete.go [hunk 2]
```diff
@@ -75,5 +75,8 @@ func deleteUser(buckets map[string]bucketWithName, uid int) (err error) {
 	if err = buckets[userByNameBucketName].Delete([]byte(u.Name)); err != nil {
 		panic(fmt.Sprintf("programming error: delete is not allowed in a RO transaction: %v", err))
 	}
+	if err = buckets[userToBrokerBucketName].Delete([]byte(strconv.Itoa(u.UID))); err != nil {
```
[the remark this answers is not in the document]
Comment (denisonbarbosa) [reply internal/cache/delete.go:78]: We do need the Itoa. The `.Delete` function is from bbolt itself and it takes a `[]bytes` as an argument, which we can't obtain directly from an int. So we need to parse it into a `string` first and then convert it to `[]bytes`.

[the remark this answers is not in the document]
Comment (jibel) [reply internal/cache/delete.go:78]: Ah true, itoa != atoi sorry

File: internal/services/pam/pam.go [hunk 3]
```diff
@@ -54,14 +54,23 @@ func (s Service) AvailableBrokers(ctx context.Context, _ *authd.Empty) (*authd.A
...
+	if b := s.brokerManager.BrokerForUser(req.GetUsername()); b != nil {
+		return &authd.GPBResponse{PreviousBroker: &b.ID}, nil
+	}
 
-	b := s.brokerManager.BrokerForUser(req.GetUsername())
-	if b != nil {
-		r.PreviousBroker = &b.ID
+	brokerID, err := s.cache.BrokerForUser(req.GetUsername())
+	if err != nil {
+		log.Warningf(ctx, "Could not get previous broker for user %q from cache: %v", req.GetUsername(), err)
```
[the remark this answers is not in the document]
Comment (denisonbarbosa) [reply internal/services/pam/pam.go]: I agree. We shouldn't already spam the user with a warning on his first login. I addressed this in the last commit now!

**Review by jibel — APPROVED** [review]
Everythings's look good. Thanks!

**Pull request merged**